### PR TITLE
adds IRule + decision model for memmapp+proglogic platform

### PR DESCRIPTION
**Pull request by BeethovenKodar**
This PR creates the IRule `MM_MCoreAndPL_IRule` and the corresponding decision model `MM_MCoreAndPL_IRule`.

Note:
- L44 in `MM_MCoreAndPL_IRule` uses the `LogicProgrammableModuleViewer` instead of `LogicProgrammableModule` as it should be, but this was decided necessary because otherwise the code has syntax errors. 
- Because @Rojods develops on Windows all newlines are added to the git changes. I'll try to fix this behavior so the actual changes are easier to identify.

### CODE REVIEW

File: java-common/src/main/java/idesyde/common/MM_MCoreAndPL.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * This type of platform is what one would expect from most COTS platforms and
+ * hardware designs, which completely or partially follows a von neumman
+ * architecture. This means that the storage elements store both data and
+ * instructions and the processors access them going through the communication
+ * elements; the latter that form the 'interconnect'.
+ */
+@JsonSerialize
+public record MM_MCoreAndPL(
+		@JsonProperty("processing_elems") Set<String> processingElems,
+		@JsonProperty("pl_module_available_areas") Map<String, Integer> plModuleSizes,
```
Comment (BeethovenKodar) [inline java-common/src/main/java/idesyde/common/MM_MCoreAndPL.java:26]: @Rojods only this line here

File: java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+		var identified = new HashSet<MM_MCoreAndPL>();
+		var errors = new HashSet<String>();
+		var model = new SystemGraph();
+		for (var dm : designModels) {
+			ForSyDeIODesignModel.tryFrom(dm).map(ForSyDeIODesignModel::systemGraph).ifPresent(model::mergeInPlace);
+		}
+		var processingElements = new HashSet<GenericProcessingModule>();
+		var memoryElements = new HashSet<GenericMemoryModule>();
+		var communicationElements = new HashSet<GenericCommunicationModule>();
+		var plElements = new HashSet<LogicProgrammableModuleViewer>(); //! TODO don't use the viewer
```
Comment (BeethovenKodar) [inline java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java:44]: @Rojods here is the initialization of the collection of PL elements

File: java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+					ForSyDeHierarchy.GenericProcessingModule
+							.tryView(model, v)
+							.ifPresent(p -> processingElements.add(p));
+					ForSyDeHierarchy.GenericMemoryModule
+							.tryView(model, v)
+							.ifPresent(p -> memoryElements.add(p));
+					ForSyDeHierarchy.GenericCommunicationModule
+							.tryView(model, v)
+							.ifPresent(p -> communicationElements.add(p));
+					ForSyDeHierarchy.LogicProgrammableModule
```
Comment (BeethovenKodar) [inline java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java:56]: @Rojods here the component(s) are added to the collection

File: java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+		// var mutMap = new HashMap<String, Map<String, Double>>();
+		// return mutMap;
+		// });
+		if (processingElements.size() <= 0) {
+			errors.add("MM_McoreAndPL_IRule: no processing elements");
+		}
+		if (memoryElements.size() <= 0) {
+			errors.add("MM_McoreAndPL_IRule: no memory elements");
+		}
+		if (plElements.size() <= 0) {
```
Comment (BeethovenKodar) [inline java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java:145]: make sure there is at least one PL element

File: java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+							ce -> ForSyDeHierarchy.InstrumentedCommunicationModule
+									.tryView(ce)
+									.stream()
+									.mapToDouble(ice -> (double) ice
+											.flitSizeInBits()
+											* (double) ice.operatingFrequencyInHertz()
+											/ (double) ice.maxCyclesPerFlit())
+									.max()
+									.orElse(0.0)));
+			var totalAvailablePLAreas = plElements.stream()
```
Comment (BeethovenKodar) [inline java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java:195]: @Rojods extract the area from each of the components that should be fed to the decision model

File: java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+					.collect(Collectors.toMap(
+							pl -> pl.getIdentifier(),
+							pl -> pl.availableLogicArea()));
+			identified.add(
+					new MM_MCoreAndPL(
+							processingElements.stream()
+									.map(x -> x.getIdentifier())
+									.collect(Collectors.toSet()),
+							totalAvailablePLAreas,
```
Comment (BeethovenKodar) [inline java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java:204]: @Rojods pass it as argument

File: java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java (hunk 4 above)
Comment (BeethovenKodar) [inline java-bridge-forsyde-io/src/main/java/idesyde/forsydeio/MM_MCoreAndPL_IRule.java:145]: @Rojods make sure there is at least one PL element

**Comment by Rojods** [comment]
It seems very cleanly additive to me, which makes me think it is an easy addition! I will merge it now and close the PR.

**Pull request merged**